### PR TITLE
Add UTM to Asset download

### PR DESCRIPTION
**Pull request by tomekkowalczyk**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          |  no
| New feature/enhancement? (use the a.x branch)      | yes
| Deprecations?                          | no
| BC breaks? (use the c.x branch)        | no
| Automated tests included?              |
| Related user documentation PR URL      | 
| Related developer documentation PR URL |
| Issue(s) addressed                     | Fixes #... 

#### Description:
The functionality allows you to register UTM tags from asset links through GET parameters. UTM columns are added to the asset download table, where the data is retrieved.

These parameters can be read from the level of asset download reports.

URL: http://mautic4.local/asset/15:test?utm_source=test2&utm_medium=test3&utm_campaign=test6&utm_id=test1&utm_term=test4&utm_content=test4

image

Report Asset Downloads
image

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/11266"></a>

### CODE REVIEW

File: app/bundles/AssetBundle/Entity/Download.php [hunk 1]
```diff
@@ -66,6 +66,32 @@ class Download
...
+
+    /**
+     * @var string
+     */
+    private $utmSource;
+
+    /**
+     * @var string
+     */
+    private $utmTerm;
```
Comment (escopecz) [inline app/bundles/AssetBundle/Entity/Download.php:92]: Let's use typed properties in new code at least. All entity properties can be null if they are not set directly through the constructor.
```suggestion
    private ?string $utmCampaign;

    private ?string $utmContent;

    private ?string $utmMedium;

    private ?string $utmSource;

    private ?string $utmTerm;
```

File: app/bundles/AssetBundle/Entity/Download.php [hunk 2]
```diff
@@ -112,6 +138,31 @@ public static function loadMetadata(ORM\ClassMetadata $metadata)
...
+            ->columnName('utm_medium')
+            ->nullable()
+            ->build();
+
+        $builder->createField('utmSource', 'string')
+            ->columnName('utm_source')
+            ->nullable()
+            ->build();
+
+        $builder->createField('utmTerm', 'string')
```
Comment (escopecz) [inline app/bundles/AssetBundle/Entity/Download.php:162]: Let's use constants over strings wherever possible
```suggestion
        $builder->createField('utmCampaign', Types::STRING)
            ->columnName('utm_campaign')
            ->nullable()
            ->build();

        $builder->createField('utmContent', Types::STRING)
            ->columnName('utm_content')
            ->nullable()
            ->build();

        $builder->createField('utmMedium', Types::STRING)
            ->columnName('utm_medium')
            ->nullable()
            ->build();

        $builder->createField('utmSource', Types::STRING)
            ->columnName('utm_source')
            ->nullable()
            ->build();

        $builder->createField('utmTerm', Types::STRING)
```

File: app/bundles/AssetBundle/Model/AssetModel.php [hunk 3]
```diff
@@ -159,6 +159,11 @@ public function trackDownload($asset, $request = null, $code = '200', $systemEnt
 
         $download = new Download();
         $download->setDateDownload(new \Datetime());
+        $download->setUtmCampaign($request->get('utm_campaign') ?? null);
+        $download->setUtmContent($request->get('utm_content') ?? null);
+        $download->setUtmMedium($request->get('utm_medium') ?? null);
+        $download->setUtmSource($request->get('utm_source') ?? null);
+        $download->setUtmTerm($request->get('utm_term') ?? null);
```
Comment (escopecz) [inline app/bundles/AssetBundle/Model/AssetModel.php:166]: Setting default value of `null` like this is not necessary as the get method returns `null` as the default value automatically.
```suggestion
        $download->setUtmCampaign($request->get('utm_campaign'));
        $download->setUtmContent($request->get('utm_content'));
        $download->setUtmMedium($request->get('utm_medium'));
        $download->setUtmSource($request->get('utm_source'));
        $download->setUtmTerm($request->get('utm_term'));
```

File: app/bundles/PageBundle/Controller/PublicController.php [hunk 4]
```diff
@@ -500,7 +500,11 @@ public function redirectAction($redirectId)
             }
 
             if (false !== strpos($url, $this->generateUrl('mautic_asset_download'))) {
-                $url .= '?ct='.$ct;
+                if (strpos($url, '&')) {
+                    $url .= '&ct=' . $ct;
+                } else {
+                    $url .= '?ct=' . $ct;
+                }
```
Comment (escopecz) [inline app/bundles/PageBundle/Controller/PublicController.php:507]: Isn't this if opposite of what it should do? If it contains `&` then append the GET param with `&`? Also, check the warning in the docs for this function: https://www.php.net/manual/en/function.strpos.php

It's better to compare with `false` as the `0` value can cause unwanted behavior.

File: app/bundles/PageBundle/Model/TrackableModel.php [hunk 5]
```diff
@@ -363,6 +363,7 @@ protected function extractTrackablesFromHtml($html)
         /** @var \DOMElement $link */
         foreach ($links as $link) {
             $url = $link->getAttribute('href');
+            $url = preg_replace('/{(\S+?)}/i', '', $url);
```
Comment (escopecz) [inline app/bundles/PageBundle/Model/TrackableModel.php:366]: What is the intent behind this replace? Do you expect whitespaces to be in the links? If so, shouldn't those be replaced with `+` or even better use https://www.php.net/manual/en/function.urlencode.php for the values in the GET params?

File: app/bundles/ReportBundle/Translations/en_US/messages.ini [hunk 6]
```diff
@@ -23,6 +23,11 @@ mautic.report.field.lead.title="Contact title"
 mautic.report.field.lead.zipcode = "Contact zip code"
 mautic.report.field.modified_by_user="Last modified by"
 mautic.report.field.publish_down="Publish down on"
+mautic.report.field.utm_campaign="UTM Campaign"
+mautic.report.field.utm_content="UTM Content"
+mautic.report.field.utm_medium="UTM Medium"
+mautic.report.field.utm_source="UTM source"
+mautic.report.field.utm_term="UTM Term"
```
Comment (escopecz) [inline app/bundles/ReportBundle/Translations/en_US/messages.ini:30]: Can you use these translation strings from the LeadBundle instead?
```
mautic.lead.report.utm.campaign="UTM campaign"
mautic.lead.report.utm.content="UTM content"
mautic.lead.report.utm.medium="UTM medium"
mautic.lead.report.utm.source="UTM source"
mautic.lead.report.utm.term="UTM term"
```
I understand that this way there is dependency of the ReportBundle on the LeadBundle but this dependency exists already in so many places that the ReportBundle cannot work without LeadBundle anyway. I think we should use what's already there and save some time for our translators.

**Review by escopecz — CHANGES_REQUESTED** [review]
Thanks for this feature! I made some comments on how to improve it. Please also take a look at failing CI jobs. I did not report the same issues that are covered by PHPSTAN like adding param and return types for the new methods.

**Comment by RCheesley** [comment]
@tomekkowalczyk it would help testers a lot if you were able to give a couple of clear steps as to how they might test this PR. I notice you already have some feedback from code review to address as well :)

**Comment by RCheesley** [comment]
In testing this PR, I was not getting any result at all in the asset downloads table despite the asset showing four downloads:
mautic ddev site_8037 _ db _ db _ assets _ phpMyAdmin 5 1 3
mautic ddev site_8037 _ db _ db _ asset_downloads _ phpMyAdmin 5 1 3

I am not sure why this is the case but would appreciate some proper testing instructions before wasting more time trying to understand myself!

**Comment by tomekkowalczyk** [comment]
@escopecz, @RCheesley 

I rewrote the solution to version 5.x and also added tests and applied the tips above https://github.com/mautic/mautic/pull/12098

**Pull request closed without merging**